### PR TITLE
Implement overridable global ID translator

### PR DESCRIPTION
**Pull request by avitex**
As per #85 

#### TODO
- [X] Documentation
- [X] Tests

### CODE REVIEW

**Comment by avitex** [comment]
LMKWYT.
Will write some documentation and a couple of tests this weekend.

File: lib/absinthe/relay/node.ex [hunk 1]
```diff
@@ -200,6 +200,9 @@ defmodule Absinthe.Relay.Node do
     end
   end
 
+  @doc """
+  Similar to `from_global_id/2` but raises `RuntimeError` if fails to decode global ID.
```
Comment (avitex) [inline lib/absinthe/relay/node.ex]: @bruce Not sure if `RuntimeError` is correct in this context (same with `from_global_id!/2`). What do you think?

File: lib/absinthe/relay/node.ex [hunk 2]
```diff
@@ -210,19 +230,57 @@ defmodule Absinthe.Relay.Node do
...
+      {:error, err} ->
+        raise RuntimeError, message: err
+    end
+  end
+
+  defp translate_global_id(nil, direction, args) do
+    apply(Absinthe.Relay.Node.IDTranslator.Default, direction, args ++ [nil])
+  end
+  defp translate_global_id(schema, direction, args) do
+    (global_id_translator(:env, schema) || global_id_translator(:schema, schema) || Absinthe.Relay.Node.IDTranslator.Default)
```
Comment (bruce) [inline lib/absinthe/relay/node.ex]: I'm pretty sure that the environment setting should be the fallback for the explicit schema (`use`) setting, not the other way around.

File: lib/absinthe/relay/schema.ex [hunk 3]
```diff
@@ -5,10 +5,26 @@ defmodule Absinthe.Relay.Schema do
   See `Absinthe.Relay`.
   """
 
-  defmacro __using__(opts) do
+  defmacro __using__ do
```
[the remark this answers is not in the document]
Comment (avitex) [reply lib/absinthe/relay/schema.ex]: Learn something new every day, never knew this.

File: lib/absinthe/relay/schema.ex [hunk 4]
```diff
@@ -5,10 +5,26 @@ defmodule Absinthe.Relay.Schema do
...
+  end
+
+  defp do_using(flavor, opts) do
     quote do
-      use Absinthe.Relay.Schema.Notation, unquote(opts)
+      use Absinthe.Relay.Schema.Notation, unquote(flavor)
       import_types Absinthe.Relay.Connection.Types
+      
+      def __absinthe_relay_global_id_translator__ do
+        Keyword.get(unquote(opts), :global_id_translator, nil)
```
[the remark this answers is not in the document]
Comment (avitex) [reply lib/absinthe/relay/schema.ex]: Yeah, this is an old habit I have in being a bit too explicit

File: test/lib/absinthe/relay/node/parse_ids_test.exs [hunk 5]
```diff
@@ -189,8 +209,8 @@ defmodule Absinthe.Relay.Node.ParseIDsTest do
 
   end
 
-  @foo1_id Base.encode64("Foo:1")
```
[the remark this answers is not in the document]
Comment (avitex) [reply test/lib/absinthe/relay/node/parse_ids_test.exs]: Yeah, I though I assume `Absinthe.Relay.NodeTest` will be enough as it tests with Base64 encoding.

[the remark this answers is not in the document]
Comment (avitex) [reply test/lib/absinthe/relay/node/parse_ids_test.exs]: Once #97 is fixed, I can refactor these tests a bit further and possibly pull the `IDTranslator` related tests to its own file.

**Comment by avitex** [comment]
In terms of backwards compatibility, I think a solution till `v1.5` could be to make `to_global_id/2` return `nil` on an error as it does now in `v1.4`.

This would mean `to_global_id!/2` would then need to return `nil` early if  `source_id` was `nil` and raise if `to_global_id/2` returns `nil` (Following the behaviour of `to_global_id(_node_type, nil, _schema)`).

Thoughts?

**Comment by bruce** [comment]
Note that `to_global_id(_node_type, nil, _schema)` doesn't `raise`, it warns and returns an error tuple.

In general, we want to avoid raising during runtime, as it doesn't give us a good method of returning useful information to the API user (exceptions are something we try to reserve for mistakes by the schema designer). 

If you can make this more backwards compatible for now, let's do it.

**Comment by avitex** [comment]
Not sure if I understand completely.

**Currently in the wild:**
- `to_global_id/2` returns [`binary | nil`](https://github.com/absinthe-graphql/absinthe_relay/blob/master/lib/absinthe/relay/node.ex#L213)
- `from_global_id/2` returns [`{:ok, nil} | {:ok, %{type: atom, id: binary}} | {:error, binary}`](https://github.com/absinthe-graphql/absinthe_relay/blob/master/lib/absinthe/relay/node.ex#L168-L169).

Perhaps you are talking about `global_id_resolver/2` which warns via [report_fetch_id_error/2](https://github.com/absinthe-graphql/absinthe_relay/blob/master/lib/absinthe/relay/node.ex#L261) and returns a error tuple?

I understand the benefits on not raising on runtime, but `to_global_id/2` currently does *not* return an error tuple, which was my reasoning behind providing `to_global_id!/2` (a stepping stone for upgrading).

I can foresee `to_global_id/2` returning error tuples on `v1.5`.

I'll make the changes I outlined in my previous comment and push them up for review.

**Comment by bruce** [comment]
Ah, yes, sorry. Apparently I misread. That's what I get for skimming through code on my phone.

Personally, I see very little benefit in using a `!` version that raises, for the reasons I mentioned earlier. I think it's fine if we provide it, but our defaults should certainly more gracefully handle errors.

**Comment by avitex** [comment]
Yeah, I agree with you. I'll remove it for this PR but we could introduce it in `v1.5` and remove in a later version. Only purpose it serves is for a quick upgrade to the next version.

**Comment by avitex** [comment]
Heya @bruce, I've removed the bang functions and fixed the conflicts. Let me know if there is anything else you'd like to discuss or change.

**Comment by tlvenn** [comment]
@bruce any chance to merge this soon ? Thanks in advance !

**Comment by madebyherzblut** [comment]
I would love to see this merged (and released) as well. We use [hashids](http://hashids.org/) to obfusicate the ID and a global ID translator would remove a lot of code. Let me know if I can help in any way 😊

**Comment by bruce** [comment]
Sorry folks, we're finally going to do some passes on PRs (we've been crushed with other stuff/exhausted) and will look at getting this in very soon.

**Comment by bruce** [comment]
This will ship in the next release (bundled with some other changes); will try to get it out today/tomorrow.

**Pull request merged**